### PR TITLE
PWX-21857: incorporate kernel 5.15 api changes

### PR DESCRIPTION
**Pull request by prabirpaul**
**What this PR does / why we need it**:
Needed to px 2.10 and master's a lot different, so I chose to try and this out on fuse v2.9.0. 
Master will have a separate PR.

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

Smoke: https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-08/34/

### CODE REVIEW

File: pxd.c [hunk 1]
```diff
@@ -859,32 +862,36 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
...
 	pxd_fastpath_cleanup(pxd_dev);
 	pxd_dev->disk = NULL;
+
+	if (disk) {
+		del_gendisk(disk);
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
-	if (disk_live(disk)) {
-#else	
-	if (disk->flags & GENHD_FL_UP) {
+		blk_cleanup_disk(disk);
```
Comment (sulakshm) [inline pxd.c:886]: blk_cleanup_disk() already does put_disk(). Likely double free here.
This cleanup path needs to be checked for differing version.

**Comment by sulakshm** [comment]
can you run a unit test of add and remove device in differing versions? 
Rest of changes seem okay to me, but not sure whether i have missed any.

File: pxd.c (hunk 1 above)
Comment (prabirpaul) [reply pxd.c:886]: Yeah I see that now. Not sure why but it worked on kernel 5.15 when i tried. Anyways updated now.

**Comment by prabirpaul** [comment]
```
[root@ip-70-0-125-162 ~]# uname -a
Linux ip-70-0-125-162.brbnca.spcsdns.net 5.15.0-0.rc7.53.fc36.x86_64 #1 SMP Tue Oct 26 20:41:43 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux                                                              
[root@ip-70-0-125-162 ~]# lsmod | grep px
px                     73728  1
[root@ip-70-0-125-162 ~]# for((i=1;i<10;i++)); do pxctl host attach v$i; done                                                                                                                        
Volume successfully attached at: /dev/pxd/pxd971805025973441106
Volume successfully attached at: /dev/pxd/pxd348210680785433118
Volume successfully attached at: /dev/pxd/pxd306133099390783697
Volume successfully attached at: /dev/pxd/pxd1102533042388331860
Volume successfully attached at: /dev/pxd/pxd1061879197161182881
Volume successfully attached at: /dev/pxd/pxd531668958931082971
Volume successfully attached at: /dev/pxd/pxd654743314907550215
Volume successfully attached at: /dev/pxd/pxd77159514952050888
Volume successfully attached at: /dev/pxd/pxd257189946302130922
[root@ip-70-0-125-162 ~]# lsmod | grep px
px                     73728  10
[root@ip-70-0-125-162 ~]# for((i=1;i<10;i++)); do pxctl host detach v$i; done                                                                                                                        
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
[root@ip-70-0-125-162 ~]# lsmod | grep px
px                     73728  1
[root@ip-70-0-125-162 ~]#
```

**Comment by prabirpaul** [comment]
```
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# uname -a
Linux prabir-VirtualBox 5.4.0-80-generic #90~18.04.1-Ubuntu SMP Tue Jul 13 19:40:02 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# 
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# lsmod | grep px
px                     61440  1
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# for((i=1;i<10;i++)); do ./bin/pxctl host attach v$i; done
Volume successfully attached at: /dev/pxd/pxd66511830077944735
Volume successfully attached at: /dev/pxd/pxd700791137577857822
Volume successfully attached at: /dev/pxd/pxd145126537946657250
Volume successfully attached at: /dev/pxd/pxd1135294815254450498
Volume successfully attached at: /dev/pxd/pxd837730696417025690
Volume successfully attached at: /dev/pxd/pxd815824760805400896
Volume successfully attached at: /dev/pxd/pxd651699174359586899
Volume successfully attached at: /dev/pxd/pxd550159321820389737
Volume successfully attached at: /dev/pxd/pxd250746348924377085
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# lsmod | grep px
px                     61440  10
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# for((i=1;i<10;i++)); do ./bin/pxctl host detach v$i; done
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
Volume successfully detached
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx# lsmod | grep px
px                     61440  1
root@prabir-VirtualBox:~/git/go/src/github.com/portworx/porx#
```

**Comment by prabirpaul** [comment]
Trying to get a bvt run with this on 5.15. That's tricky because our base container doesn't build kernel module for 5.15 so container startup can't build it locally and of course a prebuilt module is not available on yum.portworx.com.

**Comment by sulakshm** [comment]
You can manually create a directory for the kernel you choose in the
mirrors.portworx.com site.
Get creds from Jose.

Push the built kernel module in the below path.
https://mirrors.portworx.com/build-results/pxfuse/for-installer/x86_64/<uname
-a>/version/13/px.ko

Then one should be able to get atleast the kernel module for the BVT run.

File: pxd.c [hunk 2]
```diff
@@ -806,16 +780,37 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
...
 	  err = blk_mq_alloc_tag_set(&pxd_dev->tag_set);
-	  if (err)
+	  if (err) {
 		goto out_disk;
+	  }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+	  disk = blk_mq_alloc_disk(&pxd_dev->tag_set, pxd_dev);
+	  if (IS_ERR(disk)) {
+		err = PTR_ERR(disk);
```
Comment (sulakshm) [inline pxd.c:790]: disk is initialized as a error pointer. This will get freed below when control jumps to out_disk.

**Review by sulakshm — APPROVED** [review]

**Review by sulakshm — APPROVED** [review]

**Comment by zentavr** [comment]
I wonder when this code will be merged?

**Pull request closed without merging**